gvfs-helper.c

```diff
@@ -3629,7 +3629,7 @@ static enum gh__error_code do_sub_cmd__endpoint(int argc, const char **argv)
 
 	if (ec == GH__ERROR_CODE__OK)
 		printf("%s\n", data.buf);
-	else
+	else if (ec != GH__ERROR_CODE__HTTP_404)
```

Comment:
In most cases, we do care about 404s. It's just when we call `scalar clone` and target a remote repository that isn't on Azure Repos, we want to be silent about it.

Therefore, I think, we need to identify the exact call chain that leads to said unwanted error message, and ensure that the error message is suppressed _only_ in that call chain.

Reply:
> In most cases, we do care about 404s. 

That does sound quite obvious when you put it like that.

> exact call chain that leads to said unwanted error message

I'm going to look for the caller in scalar clone

> ensure that the error message is suppressed _only_ in that call chain.

and then try to pass in that flag when we want to be silent about it

> target a remote repository that isn't on Azure Repos

How would we detect that?

Reply:
Update: 

https://cs.github.com/microsoft/git/blob/34f770827353c7b5df88c8c126139da73b05c6b3/contrib/scalar/scalar.c#L663

https://github.com/microsoft/git/blob/34f770827353c7b5df88c8c126139da73b05c6b3/contrib/scalar/scalar.c#L663

I'm pretty sure this is the culprit. 

https://github.com/microsoft/git/blob/34f770827353c7b5df88c8c126139da73b05c6b3/contrib/scalar/scalar.c#L615-L625

Checking out that function:

https://github.com/microsoft/git/blob/34f770827353c7b5df88c8c126139da73b05c6b3/contrib/scalar/scalar.c#L602-L607

How do you suggest modifying this?

Reply:
Update:

I've noticed that Azure Repos URLS either start with "dev.azure.com" or have ".visualstudio.com" in the domain. 

I think `starts_with(url, "https://)` is too generous. 

We could restrict it to `starts_with(url, "https://dev.azure.com)` but that'd miss the ones with .visualstudio.com.

So we could probably split the url on `/` so that the following example URL

> https://o365exchange.visualstudio.com/O365%20Core/_git/Griffin

Would be split into

> [https: ,  , o365exchange.visualstudio.com , O365%20Core , _git , /Griffin]

Then we could check if the third entry ended with visualstudio.com. But I'm wondering if there's a simpler way to parse for that.

Reply:
You are totally correct: Since nobody but Azure Repos will even implement the GVFS protocol on the server side (the core Git project rejected the VFS for Git idea and we have partial clones instead, which in turn probably won't be supported by Azure Repos), we can _totally_ hard-code the respective host name requirement.

Instead of splitting and exact string comparisons, we could also simply use a regular expression. And we could first strip the protocol so that it is easier to validate the host name. I have something like this in mind:

```c
static int can_url_support_gvfs(const char *url)
{
	regex_t vs;
	int ret;

	if (!skip_prefix(url, "https://", &url) ||
	    (git_env_bool("GIT_TEST_ALLOW_GVFS_VIA_HTTP", 0) &&
	     skip_prefix(url, "http://", &url))
	return 0;

	/*
	 * Only Azure Repos supports the `gvfs/config` endpoint.
	 *
	 * The URLs are of the form https://dev.azure.com/... or
	 * https://<account>.visualstudio.com/...
	 */
	if (starts_with(url, "dev.azure.com/"))
		return 1;

	if (regcomp(&vs, "^[^/]*\\.visualstudio\\.com/", REG_NOSUB))
		BUG("could not compile `.visualstudio.com` regex");
	ret = regexec(&vs, url, 0, NULL, 0) != REG_NOMATCH;

	regfree(&vs);
	return ret;
} 
```

Reply:
Wait a minute... regarding hard-coding URL requirements, there is a problem with on-prem solutions: Their URLs will _not_ have dev.azure.com nor *.visualstudio.com domains.

Reply:
Okay. So what we need to do is to start a `git clone` in `gdb`, set a breakpoint on the code producing the error message, then look at the backtrace (command: `bt`) to identify the call chain, then add a flag to silence this warning, if called via the `clone` call chain.